pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/LookupTransformFunction.java

```diff
@@ -0,0 +1,284 @@
+
+    @Override
+    public String getName() {
+        return FUNCTION_NAME;
+    }
+
+    @Override
+    public void init(List<TransformFunction> arguments, Map<String, DataSource> dataSourceMap) {
+        // Check that there are correct number of arguments
+        Preconditions.checkArgument(arguments.size() >= 4,
+            "At least 4 arguments are required for LOOKUP transform function");
+        Preconditions.checkArgument(arguments.size() % 2 == 0,
+            "Should have the same number of JoinKey and JoinValue arguments");
+
+        TransformFunction dimTableNameFunction = arguments.get(0);
+        Preconditions.checkArgument(dimTableNameFunction instanceof LiteralTransformFunction,
+            "First argument must be a literal(string) representing the dimension table name");
+        _dimTableName = ((LiteralTransformFunction) dimTableNameFunction).getLiteral();
+
+        TransformFunction dimColumnFunction = arguments.get(1);
+        Preconditions.checkArgument(dimColumnFunction instanceof LiteralTransformFunction,
+            "Second argument must be a literal(string) representing the column name from dimension table to lookup");
+        _dimColumnName = ((LiteralTransformFunction) dimColumnFunction).getLiteral();
+
+        List<TransformFunction> joinArguments = arguments.subList(2,arguments.size());
+        for (int i=0; i<joinArguments.size()/2; i++) {
+            TransformFunction dimJoinKeyFunction = joinArguments.get((i*2));
+            Preconditions.checkArgument(dimJoinKeyFunction instanceof LiteralTransformFunction,
+                "JoinKey argument must be a literal(string) representing the primary key for the dimension table");
+            _joinKeys.add(((LiteralTransformFunction) dimJoinKeyFunction).getLiteral());
+
+            TransformFunction factJoinValueFunction = joinArguments.get((i*2)+1);
+            TransformResultMetadata factJoinValueFunctionResultMetadata = factJoinValueFunction.getResultMetadata();
+            Preconditions.checkArgument(factJoinValueFunctionResultMetadata.isSingleValue(),
+                "JoinValue argument must be a single value expression");
+            _joinValueFunctions.add(factJoinValueFunction);
+        }
+
+        // Validate lookup table and relevant columns
+        _dataManager = DimensionTableDataManager.getInstanceByTableName(_dimTableName + TABLE_NAME_SUFFIX);
+        Preconditions.checkArgument(_dataManager != null,
+                String.format("Dimension table does not exist: %s", _dimTableName));
+
+        _lookupColumnFieldSpec = _dataManager.getColumnFieldSpec(_dimColumnName);
+        Preconditions.checkArgument(_lookupColumnFieldSpec != null,
+                String.format("Column does not exist in dimension table: %s:%s", _dimTableName, _dimColumnName));
+
+        for (String joinKey: _joinKeys) {
+            FieldSpec pkColumnSpec = _dataManager.getColumnFieldSpec(joinKey);
+            Preconditions.checkArgument(pkColumnSpec != null,
```

Comment:
also check that all the primary keys are provided.

Reply:
I'm assuming you are referring to the validation of the `joinValue` argument below.
```
LOOKUP(tableName, columnName, joinKey, joinValue)
```
Unlike the first three arguments, `joinValue` isn't a 'literal', so its value is only resolved when we get a projection block. So rather than validating at initialization stage, we check its existence in `transformTo...Values` methods, on a row by row basis. 

For instance, take the sample query:
```sql
SELECT
    baseballStats.playerName,
    baseballStats.teamID,
    LOOKUP('baseballTeams', 'teamName', 'teamID', baseballStats.teamID)
FROM
    baseballStats
LIMIT 10
```
We can validate the first three arguments of the LOOKUP function without reading data from `baseballStats`. However the last argument is only available once we start reading table contents.

Hope this makes sense, and please let me know if I missed something.

Reply:
I meant checking the join keys are the same as the primary keys defined on a table. Assume a table has 2 columns of the primary key, then this lookup function shall take two join keys too, to ensure the same cardinality in the output.

Reply:
Got it, this makes more sense. Just updated the validation logic to assert the provided key list is equal to primary key columns. Thanks for the suggestion.